homeassistant/components/homeconnect/switch.py

```diff
@@ -0,0 +1,185 @@
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Home Connect switch."""
+
+    def get_entities():
+        """Get a list of entities."""
+        entities = []
+        data = hass.data[DOMAIN]
+        for device_dict in data.get(DEVICES, []):
+            entities += [HomeConnectPowerSwitch(device_dict["device"])]
+            entity_dicts = device_dict.get("entities", {}).get("switch", [])
+            entity_list = [HomeConnectProgramSwitch(**d) for d in entity_dicts]
+            device = device_dict["device"]
+            device.entities += entity_list
+            entities += entity_list
+        return entities
+
+    async_add_entities(await hass.async_add_executor_job(get_entities), True)
+
+
+class HomeConnectProgramSwitch(HomeConnectEntity, SwitchDevice):
+    """Switch class for Home Connect."""
+
+    def __init__(self, device, program_name):
+        """Initialize the entitiy."""
+        name = " ".join([device.appliance.name, "Program", program_name.split(".")[-1]])
+        super().__init__(device, name)
+        self.program_name = program_name
+        self._state = None
+        self._remote_allowed = None
+
+    @property
+    def is_on(self):
+        """Return true if the switch is on."""
+        return bool(self._state)
+
+    @property
+    def available(self):
+        """Return true if the entity is available."""
+        return True
+
+    def turn_on(self, **kwargs):
+        """Start the program."""
+        _LOGGER.debug("tried to turn on program %s", self.program_name)
+        try:
+            self.device.appliance.start_program(self.program_name)
+        except HomeConnectError as err:
+            _LOGGER.error("Error while trying to start program: %s", err)
+        self.async_entity_update()
+
+    def turn_off(self, **kwargs):
+        """Stop the program."""
+        _LOGGER.debug("tried to stop program %s", self.program_name)
+        try:
+            self.device.appliance.stop_program()
+        except HomeConnectError as err:
+            _LOGGER.error("Error while trying to stop program: %s", err)
+        self.async_entity_update()
+
+    def update(self):
+        """Update the switch's status."""
+        state = self.device.appliance.status.get("BSH.Common.Root.ActiveProgram", {})
+        if state.get("value", None) == self.program_name:
+            self._state = True
+        else:
+            self._state = False
+        _LOGGER.debug("Updated, new state: %s", self._state)
+
+
+def convert_to_snake(camel):
+    """Convert from CamelCase to snake_case.
+
+    Taken from https://stackoverflow.com/questions/1175208/elegant-python-function-to-convert-camelcase-to-snake-case
+    """
+    snake = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", camel)
+    return re.sub("([a-z0-9])([A-Z])", r"\1_\2", snake).lower()
+
+
+def format_key(key):
+    """Format Home Connect keys like `BSH.Something.SomeValue` to a simple `some_value`."""
+    if not isinstance(key, str):
```

Comment:
This seems weird. Why would someone call this without a string ?

Reply:
We should move that check outside of this function and put in the call location.

Reply:
I think I did this because I also apply it to the value of the JSON returned by the API (so the function name is a bit of a misnomer) to turn values like `BSH.Common.EnumType.PowerState.On` into `on`; but those values can also be numeric. And I think I wanted to have a single function in order to write a single-line dict comprehension in L185. But that is not strictly necessary of course.

Reply:
I would prefer you create a multi-line dict comprehension. Doing this check means we're swallowing errors. I rather have the code blow up when unexpected data comes in. It helps with reusability.